### PR TITLE
10941: SHOP EE Plan Match/Out-of-pocket calculator - Scenario 1

### PR DESCRIPTION
**Pull request by davidplappert**
|Field|Value|
|-----|-----|
| Redmine Ticket Number | [10941](https://devops.dchbx.org/redmine/issues/10941) |
| App-Dev Road Map # | n/a |
| Start Date | 09/20/2016 |
| Due Date | n/a  |
| App-Dev Priority | UP-1 |
| Development Story Points | 2 |
| Peer Review Story Points | 0.25 |
| Ticket Author | @sarathpyerra |

### CODE REVIEW

File: app/models/services/checkbook_services.rb [hunk 1]
```diff
@@ -0,0 +1,70 @@
+require 'nokogiri'
+
+module CheckbookServices
+  class PlanComparision
+    attr_accessor :census_employee
+    BASE_URL =  Settings.checkbook_services.base_url
```
Comment (billy-ran-away) [inline app/models/services/checkbook_services.rb]: Good job moving the URL out of the code base directly.

File: spec/models/census_employee_spec.rb [hunk 2]
```diff
@@ -808,6 +808,50 @@
...
+      allow(ApplicationEventKind).to receive_message_chain(:where,:first).and_return(double("ApplicationEventKind",{:notice_triggers => notice_triggers,:title => "title"}))
+      allow_any_instance_of(CheckbookServices::PlanComparision).to receive(:generate_url).and_return("http://temp.url")
+    end
+    context "#generate_and_deliver_checkbook_url" do 
+      it "should create a builder and deliver without expection" do
+        expect{census_employee.generate_and_deliver_checkbook_url}.not_to raise_error
+      end
+     
+      it 'should trigger deliver' do 
+        expect(builder).to receive(:deliver)
```
Comment (billy-ran-away) [inline spec/models/census_employee_spec.rb]: Great job expanding the specs and adding more expectations.

File: app/controllers/employers/employer_profiles_controller.rb [hunk 3]
```diff
@@ -1,7 +1,8 @@
 class Employers::EmployerProfilesController < Employers::EmployersController
 
   before_action :find_employer, only: [:show, :show_profile, :destroy, :inbox,
-                                       :bulk_employee_upload, :bulk_employee_upload_form, :download_invoice, :export_census_employees, :link_from_quote]
+
```
Comment (Kevin-Wei) [inline app/controllers/employers/employer_profiles_controller.rb]: @sarathpyerra spacing

File: app/jobs/outofpocket_notifier_job.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    Resque.logger.level = Logger::DEBUG
+    ep=EmployerProfile.find(employer_profile_id)
+    event_kind = ApplicationEventKind.where(:event_name => 'out_of_pocker_url_notifier').first
+    notice_trigger = event_kind.notice_triggers.first
+    builder = notice_trigger.notice_builder.camelize.constantize.new(ep, {
+              template: notice_trigger.notice_template,
+              subject: event_kind.title,
+              mpi_indicator: notice_trigger.mpi_indicator,
+              }.merge(notice_trigger.notice_trigger_element_group.notice_peferences)).deliver
+
```
Comment (Kevin-Wei) [inline app/jobs/outofpocket_notifier_job.rb]: @sarathpyerra spacing

File: app/models/census_employee.rb [hunk 5]
```diff
@@ -48,6 +50,7 @@ class CensusEmployee < CensusMember
   before_save :assign_default_benefit_package
   before_save :allow_nil_ssn_updates_dependents
 
+
```
Comment (Kevin-Wei) [inline app/models/census_employee.rb]: @sarathpyerra spacing

File: app/models/employer_profile.rb [hunk 6]
```diff
@@ -818,6 +818,18 @@ def is_conversion?
...
+  def generate_and_deliver_checkbook_urls_for_employees
+    census_employees.each do |census_employee|
+      census_employee.generate_and_deliver_checkbook_url
+    end
+  end
+
+  def generate_checkbook_notices
+    OutofpocketNotifierJob.perform_later(self.id.to_s)
+  end
+
```
Comment (Kevin-Wei) [inline app/models/employer_profile.rb]: @sarathpyerra spacing

File: app/models/employer_profile.rb [hunk 7]
```diff
@@ -818,6 +818,18 @@ def is_conversion?
...
+
```
Comment (Kevin-Wei) [inline app/models/employer_profile.rb]: @sarathpyerra spacing

File: app/models/employer_profile.rb [hunk 8]
```diff
@@ -818,6 +818,18 @@ def is_conversion?
...
+
```
Comment (Kevin-Wei) [inline app/models/employer_profile.rb]: @sarathpyerra spacing

File: app/models/family.rb [hunk 9]
```diff
@@ -243,7 +243,7 @@ def is_under_shop_open_enrollment?
   def current_eligible_open_enrollments
     current_shop_eligible_open_enrollments + current_ivl_eligible_open_enrollments
   end
-
+
```
Comment (Kevin-Wei) [inline app/models/family.rb]: @sarathpyerra space!

File: app/models/services/checkbook_services.rb [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+      if uri.present?
+        return BASE_URL+uri.first.value
+      else
+        raise "Unable to generate url"
+      end
+      rescue Exception => e
+        puts "Exception #{e}"
+      end
+    end
+
```
Comment (Kevin-Wei) [inline app/models/services/checkbook_services.rb]: @sarathpyerra spacing

File: app/models/services/checkbook_services.rb [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+      premium_benefit_contributions = {}
+      census_employee.employer_profile.plan_years.first.benefit_groups.first.relationship_benefits.each do |relationship_benefit| 
+        next if ["child_26_and_over","nephew_or_niece","grandchild","child_26_and_over_with_disability"].include? relationship_benefit.relationship
+        relationship=  relationship_benefit.relationship == "child_under_26" ? "child" : relationship_benefit.relationship
+        premium_benefit_contributions[relationship] = relationship_benefit.premium_pct.to_f
+      end
+      premium_benefit_contributions
+    end
+
+
```
Comment (Kevin-Wei) [inline app/models/services/checkbook_services.rb]: @sarathpyerra spacing

File: app/views/employers/employer_profiles/my_account/_census_employees.html.erb [hunk 12]
```diff
@@ -124,3 +131,21 @@
   </div>
 </div>
 <div></div>
+
```
Comment (Kevin-Wei) [inline app/views/employers/employer_profiles/my_account/_census_employees.html.erb]: @sarathpyerra spacing

File: config/settings.yml [hunk 13]
```diff
@@ -95,6 +95,11 @@ aca:
       maximum_length:
         months: 2
 paper_notice: 'paper-notices'
+checkbook_services: 
+    remote_access_key: 'B48E5D58B6A64B3E93A6BF719647E568'
+    base_url:  "https://staging.checkbookhealth.org"
+    url: "https://staging.checkbookhealth.org/shop/dc/"
+
```
Comment (Kevin-Wei) [inline config/settings.yml]: @sarathpyerra space!

File: spec/factories/notice_triggers.rb [hunk 14]
```diff
@@ -1,5 +1,12 @@
 FactoryGirl.define do
   factory :notice_trigger do
-    
+
```
Comment (Kevin-Wei) [inline spec/factories/notice_triggers.rb]: @sarathpyerra spacing

File: spec/models/census_employee_spec.rb [hunk 15]
```diff
@@ -854,6 +854,50 @@
...
+        expect{census_employee.generate_and_save_to_temp_folder}.not_to raise_error
+      end
+
+       it 'should not trigger deliver' do 
+        expect(builder).not_to receive(:deliver)
+        census_employee.generate_and_save_to_temp_folder
+      end
+    end
+  end
+
```
Comment (Kevin-Wei) [inline spec/models/census_employee_spec.rb]: @sarathpyerra spacing

File: spec/models/plan_year_spec.rb [hunk 16]
```diff
@@ -25,9 +25,10 @@
       fte_count: valid_fte_count
     }
   end
-
+
```
Comment (Kevin-Wei) [inline spec/models/plan_year_spec.rb]: @sarathpyerra space!

**Comment by lokelokesh** [comment]
This was merged to master, sha - 1b05a7105d2915b5a330a758f5bcd21b615fe669

**Pull request closed without merging**